### DISCUSSION TITLE
Can apiFetch (not) be used on remote sites REST API?

### DISCUSSION THREAD

**Discussion by hansschuijff** [WordPress/gutenberg#46780] [Q&A]
Is it possible to use apiFetch on remote urls? I have a use case for that, but wasn't able to get it to request to any other website than the one I was on.

I maintain multiple websites with events and have been trying out apiFetch to make a Remote tickets block that retrieves tickets info via REST from one website to show on another website. But I have found no way to change the rootURL. I've tried passing the url option instead of the path option and I've tried the code that is offered in de docs for the rootURL middleware.

In both cases the call was performed to the domain of the website I am on and I wonder if there's a limitation in apiFetch so it only fetches from the samesite and not from a remote site.

I was easily able to do what I wanted using window.fetch, axios and XMLHTMLrequest, but not with apiFetch. And as I understand it apiFetch by default uses window.fetch for it's requests.

I haven't found any working example of apiFetch being used on a remote website's REST API, nor did I found any explicit info that states whether it's possible )or not), so I'm wondering if it's even possible/allowed. I've seen no effect on the request url by using the createRootURLMiddleware, so it doesn't seem to work for me anyway. All examples and tutorials use path and stay on the samesite. This block shouldn't.

The code below run from a plugin on a local (by Flywheel) install just resulted in a call to https://sandbox.local/wp-json/wp/v2/posts/
Latest wp version and firefox dev browser.

```
import apiFetch from '@wordpress/api-fetch';
const rootURL = 'https://dewitteprins.nl/wp-json';
apiFetch.use( apiFetch.createRootURLMiddleware( rootURL ) );
apiFetch( {
	path: '/wp/v2/posts/'
} )
.then( ( posts ) => {
	console.log( "apiFetch result: ", posts );
} ).catch( ( err ) => {
	console.log( "apiFetch error: ", err );
});
```

Anyone of you know if this should work, that remote site requests should be possible? Or is it indeed just for samesite requests?

And in the latter case, why does the doc speak about the url option for passing a complete URL and createRootURLMiddleware to change the rootURL of the request? It doesn't say the api is only for same site urls. If so, perhaps it should.

Can you help me out with this? Much appreciated.

**Comment by t-hamano** [comment]
Hi @hansschuijff,

`apiFetch` has a [url](https://github.com/WordPress/gutenberg/tree/trunk/packages/api-fetch#url-string) property, can this be used?

```javascript
wp.apiFetch( {
	url: 'https://yourapi.com',
	method: 'GET',
} ).then( ( res ) => {
	console.log( res );
} );
```

**Reply by hansschuijff** [reply]
Hi @t-hamano thanks for helping me!

I had tried the url option first, but had not included that code in my question. It didn't gave me the result I looked for, but your reaction inspired me to try again an I found that it actually performs the get and receives a 403 response.

My code:

```
import apiFetch from '@wordpress/api-fetch';
apiFetch( {
	url: 'https://dewitteprins.nl/wp-json/wp/v2/posts',
	method: 'GET',
} ).then( ( res ) => {
	console.log( "apiFetch result: ", res );
} ).catch( ( err ) => {
	console.log( "apiFetch error: ", err );
} );

```

In the console it says:
`XHR GET https://dewitteprins.nl/wp-json/wp/v2/posts?_locale=user   [HTTP/3 403 Forbidden 2676ms]`

The 403 response seems strange since the same request, when performed by just putting the address to the browser gives me the json output with no problem and also fetch() just works for this input. So somehow apifetch adds something to the request that makes the cross site requests fail by default. It has no problem with samesite requests.

When I open the above mentioned XHR (error) row in console, it says in the response tab:

```
JSON
code	"rest_cookie_invalid_nonce"
message	"Cookie controle mislukt"
data	Object { status: 403 }
status	403

```

And the Headers tab says:

```
Status: 403 Forbidden
Versie: HTTP/3
Overgebracht: 1,24 kB (grootte 94 B)
Referrer-beleid: strict-origin-when-cross-origin

```

RESPONSE-HEADER:
```
access-control-allow-credentials: true
access-control-allow-headers: Authorization, X-WP-Nonce, Content-Disposition, Content-MD5, Content-Type
access-control-allow-methods: OPTIONS, GET, POST, PUT, PATCH, DELETE
access-control-allow-origin: https://sandbox.local
access-control-expose-headers: X-WP-Total, X-WP-TotalPages, Link
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
cf-cache-status: DYNAMIC
cf-ray: 77f0ad13fea27138-YUL
content-encoding: br
content-type: application/json; charset=UTF-8
date: Sun, 25 Dec 2022 09:51:06 GMT
link: <https://dewitteprins.nl/wp-json/>; rel="https://api.w.org/"
nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=QG65CPklJc7%2FKhKIA8qkqLRQM%2Fzj8%2BbCpJKB4H7Gz%2BFhaw7IIiJqhWL%2Bq26l5wqGSaPIT%2FNpnSJo5LPPVpO6EMU3w5NMx8r6CChgWtIWAJFLSiPMH3iSLf9GU%2FObdcWJejA%3D"}],"group":"cf-nel","max_age":604800}
server: cloudflare
vary: Accept-Encoding
vary: Origin
x-content-type-options: nosniff
x-content-type-options: nosniff
x-powered-by: centminmod
x-robots-tag: noindex
x-xss-protection: 1; mode=block

```

REQUEST-HEADER:
```
Accept: application/json, */*;q=0.1
Accept-Encoding: gzip, deflate, br
Accept-Language: nl,en-US;q=0.8,de;q=0.5,en;q=0.3
Alt-Used: dewitteprins.nl
Cache-Control: no-cache
Connection: keep-alive
Host: dewitteprins.nl
Origin: https://sandbox.local
Pragma: no-cache
Referer: https://sandbox.local/
Sec-Fetch-Dest: empty
Sec-Fetch-Mode: cors
Sec-Fetch-Site: cross-site
TE: trailers
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/109.0
X-WP-Nonce: 342c078fcf

```

The main difference when I found a succesfull request using fetch(), is that the response header says:  `allow: GET`  and in the request header it says `Accept: */*` and `X-WP-Nonce: 342c078fcf`, the rest seems basicly the same. 

Since Allow: GET is in the response header it might not be of much relevance and since the response of a REST call would be json, the difference in Accept: might not be deciding. That leaves the nonce. So can it be the nonce that is added by apiFetch that is breaking cross site GET requests? Can I add permission to the get so it will work?

Don't know if I can influence the header directly using apiFetch, so I can try some more. But it still fails for now.

I did try to add the nonce middleware example from the docs too, but that didn't change the outcome.

Any suggestions?

regards,

Hans

**Reply by t-hamano** [reply]
I am not that familiar with apiFetch, but you may need the `headers` property to override the default headers.

```javascript
wp.apiFetch( {
	url: 'https://yourapi.com',
	method: 'GET',
	headers: {
		...
	}
} ).then( ( res ) => {
	console.log( res );
} );
```

**Reply by hansschuijff** [reply]
Thank you, @t-hamano, for helping. I tried your suggestion, but it didn't solve my challenge yet. 

The problem I'm facing seems to be that the nonce that is being passed fails. I've been able to find the line in the code that gets the nonce and it is a request to {local domain}/wp-admin/admin-ajax.php?action=rest-nonce. 

That tells me the nonce is generated by the local site and that works only when the request is made to the same site. The cron test will fail when the request is send to the REST API of a remote site (that doesn't even have to be a WordPress website). 

Or am I missing something?

To test this I need a way to remove the x-wp-nonce key from the request header. I already know that the other fetch options (window.fetch, XMLHTMLrequest, axios) don't add a nonce and give me the result I want. Right now it seems apiFetch only performs on samesite and is hardwired to fail on any other fetch, but if that is intentional, perhaps the docs should at least say so.

The headers option can only add and override header keys, but doesn't seem to be able to remove them and the code is setup to add x-wp-nonce to the header whenever it's not passed in the headers option. It initially runs the nonce middleware for that.

I tried the same call on other wordpress websites, to make sure caching or config isn't the problem, but that other websites also get the same error response. It is clear to me now the nonce is the problem here and it shouldn't be passed here.

That is how far I'm now. I was not yet able to make it work. What I need is a way to block the nonce to be added to the request header. It might need a change in the core code.

**Reply by hansschuijff** [reply]
I added an issue  #46787 for this limitation, since it seems part of WP's core behaviour. Thanks for your inputs.

**Reply by t-hamano** [reply]
@hansschuijff
Thank you for submitting the issue!
AFAIK I think the only way to retrieve data from a remote site's REST API is either:

1. Use `window.fetch` directly
2. Allow `X-WP-Nonce` headers on REST API servers

I think the root problem in this case is that the request header using `apiFetch` always contains `X-WP-Nonce`. The error is probably due to the fact that this header is not allowed on the remote site.

Therefore, I think the second approach might be also valid.

**Reply by hansschuijff** [reply]
@t-hamano I agree. 

The second  point, however, cannot be the solution, since both samesite requests as remote requests target a wordpress install. It's not that the other server doesn't allow x-wp-nonce headers. And solutions that are dependent on how the remote server is configured are by definition limited, since you only control the server you're on.

It's an issue now for the gutenberg team, so perhaps they will think of something more clever. Otherwise I know I just have to use window.fetch.

**Reply by hansschuijff** [reply]
By the way, since apiFetch is by default basically a wrapper around windows.fetch, it should be logical to support all requests that fetch enables. That would make it a more logical choice for blocks. Having to switch from apifetch to fetch for this reason just makes it more likely that apifetch will be skipped altogether. So IMO there must be either an apifetch strategy for supporting remote requests or a more specific failure that makes it crystal clear that apiFetch only supports samesite requests. Since it seems that all that is needed is a way to give control to devs over the auto-adding of a nonce, I would imagine supporting that would make the most sense here. But I will wait for a response on the issue for that.
